contracts/Escrow_V3.sol

```diff
@@ -0,0 +1,149 @@
+pragma solidity 0.5.1;
+
+import "./ECTools.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+/**
+ * @title Escrow_V3
+ * @dev Escrow_V3 is the latest version of the escrow contract, currently being used for production
+ */
+contract Escrow_V3 {
+    using SafeMath for uint256;
+
+    ERC20 public tokenContract;
+
+    mapping (address => bool) public signers;
+    mapping (uint256 => bool) public usedNonces;
+
+    address payable public dAppAdmin;
+    uint256 constant public REFUNDING_LOGIC_GAS_COST = 7901; // gas used for single refund
+
+    uint256 constant public FIAT_PAYMENT_FUND_FUNCTION_CALL_GAS_USED = 32831; // approximated gas for calling fundForFiatPayment
+    uint256 constant public RELAYED_PAYMENT_FUND_FUNCTION_CALL_GAS_USED = 32323; // approximated gas for calling fundForRelayedPayment
+
+    /**
+    * @dev Restricts the access to a given function to the dApp admin only
+    */
+    modifier onlyDAppAdmin() {
+        require(msg.sender == dAppAdmin, "Unauthorized access");
+        _;
+    }
+
+    /**
+    * @dev Checks whether the nonce in the authorisation signature was already used. Prevents replay attacks.
+    */
+    modifier preValidateFund(uint256 nonce) {
+        require(!usedNonces[nonce], "Nonce already used");
+        _;
+    }
+
+    /**
+    * @dev The token address and the dappadmin are set on contract creation
+    */
+    constructor(address tokenAddress, address payable _dAppAdmin) public {
+        dAppAdmin = _dAppAdmin;
+        tokenContract = ERC20(tokenAddress);
+    }
+   
+    /**
+    * @dev Funds the `addressToFund` with the proided `weiAmount`
+    * Signature from the dapp is used in order to authorize the funding
+    * The msg sender is refunded for the transaction costs
+    */
+    function fundForRelayedPayment(
+        uint256 nonce,
+        uint256 gasprice,
+        address payable addressToFund,
+        uint256 weiAmount,
+        bytes memory authorizationSignature) public preValidateFund(nonce)
+    {
+        uint256 gasLimit = gasleft().add(RELAYED_PAYMENT_FUND_FUNCTION_CALL_GAS_USED);
+
+        bytes32 hashedParameters = keccak256(abi.encodePacked(nonce, address(this), gasprice, addressToFund, weiAmount));
+        _preFund(hashedParameters, authorizationSignature, nonce);
+
+        addressToFund.transfer(weiAmount);
+
+        _refundMsgSender(gasLimit, gasprice);
+    }
+
+    /**
+    * @dev Funds the `addressToFund` with the proided `weiAmount` and `tokenAmount`
+    * Signature from the dapp is used in order to authorize the funding
+    * The msg sender is refunded for the transaction costs
+    */
+    function fundForFiatPayment(
```

Comment:
In my opinion it is better to handle the problem in the backend by using a more up-to-date version of the Gas station API.

The bare minimum gas price is 1 Gwei enforced by the protocol. If we set the minimum gas price to another value, what should this value be? 
If we set it to 5 gwei for example, we will be paying premium in gas price in times where the network has 1-5Gwei gasprice. 

Regarding the `transaction won't be processed from miners` - the transaction will be in the mempool unless some miner picks it up, so yeah it will take more time, but i think that we need to regulate this from the backend and not the contract